modules/schema-util/src/main/scala/edu/gemini/schema/util/SchemaStitcher.scala

```diff
@@ -0,0 +1,177 @@
+        node.v,
+        node.src.mkString("\n"),
+        els
+      )
+
+    def merge(l: List[SchemaNode], acc: List[SchemaNode]): List[SchemaNode] =
+      l match {
+        case Nil      => acc
+        case x :: Nil => acc :+ x
+        case x :: xx  =>
+          val (m, r) = xx.partition(_.name === x.name)
+          val s      = m.fold(x)((a, b) => SchemaNode(a.name, a.src, a.elements.union(b.elements)))
+          merge(r, acc :+ s)
+      }
+  }
+
+  sealed trait Elements
+  case object AllElements                                      extends Elements
+  final case class ElementList(l: NonEmptySet[NonEmptyString]) extends Elements
+
+  extension (el: Elements) {
+    def union(other: Elements): Elements = (el, other) match {
+      case (ElementList(l1), ElementList(l2)) => ElementList(l1 ++ l2)
+      case _                                  => AllElements
+    }
+  }
+
+  case class DependencyNode(
+    v:            Path,
+    src:          List[String],
+    dependencies: List[(Elements, DependencyNode)]
+  )
+
+  case class SchemaNode(name: Path, src: String, elements: Elements) {
+    def asString: String = elements match {
+      case AllElements    => src
+      case ElementList(l) => asString(l)
+    }
+
+    def asString(l: NonEmptySet[NonEmptyString]): String =
+      Schema(src) match {
+        case Ior.Right(b)   =>
+          resolveTypes(b.types,
+                       l.toList.map(x => b.types.find(_.name === x.toString)).flattenOption,
+                       List.empty
+          )
+            .map(SchemaRenderer.renderTypeDefn)
+            .mkString("\n")
+        case Ior.Both(_, b) =>
+          resolveTypes(b.types,
+                       l.toList.map(x => b.types.find(_.name === x.toString)).flattenOption,
+                       List.empty
+          )
+            .map(SchemaRenderer.renderTypeDefn)
+            .mkString("\n")
+        case _              => ""
+      }
+
+    def resolveTypes(
+      types:    List[NamedType],
+      newNames: List[NamedType],
+      acc:      List[NamedType]
+    ): List[NamedType] =
+      if (newNames.isEmpty) {
+        acc
+      } else {
+        val uniqueNews = newNames.distinct
+        val nextVals   = uniqueNews
+          .flatMap {
+            case fields: TypeWithFields                => fields.fields.map(_.tpe.underlying.asNamed).flattenOption
+            case UnionType(name, description, members) => members
+            case _                                     => List.empty
+          }
+          .map(_.dealias)
+          .filter {
+            case u: ScalarType => !u.isBuiltIn
+            case v             => !acc.contains(v) && !uniqueNews.contains(v)
+          }
+        resolveTypes(types, nextVals, uniqueNews ++ acc)
+      }
+  }
+
+  given Order[NonEmptyString] = Order.by(_.toString)
+
+  val allElementsParser: Parser[Elements]                        = char('*').as(AllElements)
```

Comment:
I guess these could all be private. Being public makes them part of the mima signature thus harder to change in the future